File: ios/LookSeeProto/Models/PromotionService.swift
//
//  PromotionService.swift
//  LookSeeProto
//

import Foundation
import Combine

// MARK: - Promotion API Model
// Separate from the SwiftUI Promotion struct — this is what goes over the wire.

struct PromotionPayload: Identifiable, Codable {
    let id: String           // promotionId (UUID string)
    let userEmail: String
    let landmarkId: String
    let landmarkLabel: String
    let name: String
    let description: String
    let startDate: String    // ISO 8601: "2026-04-21"
    let endDate: String
    let enabled: Bool
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id = "promotionId"
        case userEmail, landmarkId, landmarkLabel
        case name, description, startDate, endDate, enabled, createdAt
    }
}

// MARK: - Request bodies

private struct CreatePromotionRequest: Encodable {
    let userEmail: String
    let landmarkId: String
    let landmarkLabel: String
    let name: String
    let description: String
    let startDate: String
    let endDate: String
    let enabled: Bool
}

private struct UpdatePromotionRequest: Encodable {
    let userEmail: String
    let landmarkId: String
    let landmarkLabel: String
    let name: String
    let description: String
    let startDate: String
    let endDate: String
    let enabled: Bool
}

// MARK: - Service

@MainActor
final class PromotionService: ObservableObject {
    @Published var promotions: [PromotionPayload] = []
    @Published var isLoading = false
    @Published var errorMessage: String? = nil

    private let baseURL = URL(string: "https://7gmn5z3uf2.execute-api.us-east-1.amazonaws.com/dev")!

    private static let isoFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    // MARK: - GET /promotions?userEmail=...
    func fetchPromotions(userEmail: String) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        var components = URLComponents(
            url: baseURL.appendingPathComponent("promotions"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [URLQueryItem(name: "userEmail", value: userEmail)]

        guard let url = components.url else {
            errorMessage = "Invalid URL."
            return
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode) else {
                errorMessage = "Failed to load promotions."
                return
            }
            let decoded = try JSONDecoder().decode(PromotionListResponse.self, from: data)
            promotions = decoded.items
        } catch {
            errorMessage = "Failed to load promotions: \(error.localizedDescription)"
        }
    }

    // MARK: - POST /promotions
    func createPromotion(
        userEmail: String,
        landmarkId: String,
        landmarkLabel: String,
        name: String,
        description: String,
        startDate: Date,
        endDate: Date,
        enabled: Bool = true
    ) async {
        errorMessage = nil

        let body = CreatePromotionRequest(
            userEmail: userEmail,
            landmarkId: landmarkId,
            landmarkLabel: landmarkLabel,
            name: name,
            description: description,
            startDate: Self.isoFormatter.string(from: startDate),
            endDate: Self.isoFormatter.string(from: endDate),
            enabled: enabled
        )

        do {
            let created = try await post(
                path: "promotions",
                body: body,
                responseType: PromotionPayload.self
            )
            promotions.append(created)
        } catch {
            errorMessage = "Failed to create promotion: \(error.localizedDescription)"
        }
    }

    // MARK: - PATCH /promotions/{promotionId}
    func updatePromotion(
        promotionId: String,
        userEmail: String,
        landmarkId: String,
        landmarkLabel: String,
        name: String,
        description: String,
        startDate: Date,
        endDate: Date,
        enabled: Bool
    ) async {
        errorMessage = nil

        let body = UpdatePromotionRequest(
            userEmail: userEmail,
            landmarkId: landmarkId,
            landmarkLabel: landmarkLabel,
            name: name,
            description: description,
            startDate: Self.isoFormatter.string(from: startDate),
            endDate: Self.isoFormatter.string(from: endDate),
            enabled: enabled
        )

        let url = baseURL.appendingPathComponent("promotions/\(promotionId)")

        do {
            var req = URLRequest(url: url)
            req.httpMethod = "PATCH"
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(body)

            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode) else {
                errorMessage = "Failed to update promotion."
                return
            }
            let updated = try JSONDecoder().decode(PromotionPayload.self, from: data)
            if let index = promotions.firstIndex(where: { $0.id == promotionId }) {
                promotions[index] = updated
            }
        } catch {
            errorMessage = "Failed to update promotion: \(error.localizedDescription)"
        }
    }

    // MARK: - DELETE /promotions/{promotionId}?userEmail=...
    func deletePromotion(promotionId: String, userEmail: String) async {
        errorMessage = nil

        var components = URLComponents(
            url: baseURL.appendingPathComponent("promotions/\(promotionId)"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [URLQueryItem(name: "userEmail", value: userEmail)]

        guard let url = components.url else {
            errorMessage = "Invalid URL."
            return
        }

        do {
            var req = URLRequest(url: url)
            req.httpMethod = "DELETE"

            let (_, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode) else {
                errorMessage = "Failed to delete promotion."
                return
            }
            promotions.removeAll { $0.id == promotionId }
        } catch {
            errorMessage = "Failed to delete promotion: \(error.localizedDescription)"
        }
    }

    // MARK: - Generic POST helper

    private func post<Body: Encodable, Response: Decodable>(
        path: String,
        body: Body,
        responseType: Response.Type
    ) async throws -> Response {
        let url = baseURL.appendingPathComponent(path)
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse,
              (200...299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ServiceError.badStatus(body)
        }
        return try JSONDecoder().decode(responseType, from: data)
    }
}

// MARK: - Helpers

private struct PromotionListResponse: Decodable {
    let items: [PromotionPayload]
}

private enum ServiceError: LocalizedError {
    case badStatus(String)
    var errorDescription: String? {
        if case .badStatus(let msg) = self { return msg }
        return nil
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/PromotionService.kt
package looksee.angelll.com.models

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class PromotionPayload(
    @SerializedName("promotionId")
    val id: String = "",
    val userEmail: String = "",
    val landmarkId: String = "",
    val landmarkLabel: String = "",
    val name: String = "",
    val description: String = "",
    val startDate: String = "",
    val endDate: String = "",
    val enabled: Boolean = false,
    val createdAt: String? = null,
)

private data class PromotionListResponse(
    val items: List<PromotionPayload> = emptyList(),
)

private data class PromotionMutationRequest(
    val userEmail: String,
    val landmarkId: String,
    val landmarkLabel: String,
    val name: String,
    val description: String,
    val startDate: String,
    val endDate: String,
    val enabled: Boolean,
)

class PromotionServiceException(
    val statusCode: Int,
    val responseBody: String,
) : Exception("Promotion API error $statusCode: $responseBody")

/** Legacy/public promotion API retained separately from BusinessPromotionService. */
class PromotionService internal constructor(
    private val httpClient: BusinessHttpClient,
    private val gson: Gson,
) {
    constructor() : this(UrlConnectionBusinessHttpClient(), Gson())

    internal constructor(httpClient: BusinessHttpClient) : this(httpClient, Gson())

    private val _promotions = MutableStateFlow<List<PromotionPayload>>(emptyList())
    val promotions: StateFlow<List<PromotionPayload>> = _promotions.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    suspend fun fetchPromotions(userEmail: String) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            _promotions.value = fetchList(
                urlWithQuery(
                    "$LOOKSEE_API_BASE_URL/promotions",
                    mapOf("userEmail" to userEmail),
                ),
            )
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            _errorMessage.value = "Failed to load promotions: " +
                (error.message ?: "Unknown error")
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun fetchPromotionsForLandmark(landmarkId: String): List<PromotionPayload> =
        fetchListOrEmpty(
            urlWithQuery(
                "$LOOKSEE_API_BASE_URL/promotions/by-landmark",
                mapOf("landmarkId" to landmarkId),
            ),
        )

    suspend fun fetchPromotionsByLabel(label: String): List<PromotionPayload> =
        fetchListOrEmpty(
            urlWithQuery(
                "$LOOKSEE_API_BASE_URL/promotions/by-label",
                mapOf("landmarkLabel" to label),
            ),
        )

    suspend fun createPromotion(
        userEmail: String,
        landmarkId: String,
        landmarkLabel: String,
        name: String,
        description: String,
        startDate: LocalDate,
        endDate: LocalDate,
        enabled: Boolean = true,
    ) {
        _errorMessage.value = null
        try {
            val created = mutate(
                method = "POST",
                url = "$LOOKSEE_API_BASE_URL/promotions",
                body = mutationBody(
                    userEmail,
                    landmarkId,
                    landmarkLabel,
                    name,
                    description,
                    startDate,
                    endDate,
                    enabled,
                ),
            )
            _promotions.value = _promotions.value + created
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            _errorMessage.value = "Failed to create promotion: " +
                (error.message ?: "Unknown error")
        }
    }

    suspend fun updatePromotion(
        promotionId: String,
        userEmail: String,
        landmarkId: String,
        landmarkLabel: String,
        name: String,
        description: String,
        startDate: LocalDate,
        endDate: LocalDate,
        enabled: Boolean,
    ) {
        _errorMessage.value = null
        try {
            val updated = mutate(
                method = "PATCH",
                url = "$LOOKSEE_API_BASE_URL/promotions/${encodedPathSegment(promotionId)}",
                body = mutationBody(
                    userEmail,
                    landmarkId,
                    landmarkLabel,
                    name,
                    description,
                    startDate,
                    endDate,
                    enabled,
                ),
            )
            _promotions.value = _promotions.value.map { current ->
                if (current.id == promotionId) updated else current
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            _errorMessage.value = "Failed to update promotion: " +
                (error.message ?: "Unknown error")
        }
    }

    suspend fun deletePromotion(promotionId: String, userEmail: String) {
        _errorMessage.value = null
        try {
            execute(
                BusinessHttpRequest(
                    method = "DELETE",
                    url = urlWithQuery(
                        "$LOOKSEE_API_BASE_URL/promotions/" +
                            encodedPathSegment(promotionId),
                        mapOf("userEmail" to userEmail),
                    ),
                ),
            )
            _promotions.value = _promotions.value.filterNot { it.id == promotionId }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            _errorMessage.value = "Failed to delete promotion: " +
                (error.message ?: "Unknown error")
        }
    }

    private suspend fun fetchList(url: String): List<PromotionPayload> {
        val response = execute(BusinessHttpRequest(method = "GET", url = url))
        return gson.fromJson(response.bodyText, PromotionListResponse::class.java)?.items
            ?: emptyList()
    }

    private suspend fun fetchListOrEmpty(url: String): List<PromotionPayload> = try {
        fetchList(url)
    } catch (error: CancellationException) {
        throw error
    } catch (_: Throwable) {
        emptyList()
    }

    private suspend fun mutate(
        method: String,
        url: String,
        body: PromotionMutationRequest,
    ): PromotionPayload {
        val response = execute(
            BusinessHttpRequest(
                method = method,
                url = url,
                body = gson.toJson(body).toByteArray(Charsets.UTF_8),
                contentType = "application/json",
            ),
        )
        return gson.fromJson(response.bodyText, PromotionPayload::class.java)
            ?: throw IllegalStateException("The promotion response was empty.")
    }

    private suspend fun execute(request: BusinessHttpRequest): BusinessHttpResponse {
        val response = httpClient.execute(request)
        if (response.statusCode !in 200..299) {
            throw PromotionServiceException(response.statusCode, response.bodyText)
        }
        return response
    }

    private fun mutationBody(
        userEmail: String,
        landmarkId: String,
        landmarkLabel: String,
        name: String,
        description: String,
        startDate: LocalDate,
        endDate: LocalDate,
        enabled: Boolean,
    ) = PromotionMutationRequest(
        userEmail = userEmail,
        landmarkId = landmarkId,
        landmarkLabel = landmarkLabel,
        name = name,
        description = description,
        startDate = DATE_FORMATTER.format(startDate),
        endDate = DATE_FORMATTER.format(endDate),
        enabled = enabled,
    )

    private companion object {
        val DATE_FORMATTER: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE
    }
}
